src/systems/pose_publisher/PosePublisher.cc

```diff
@@ -155,16 +155,6 @@ class gz::sim::systems::PosePublisherPrivate
   /// improves performance by avoiding memory allocation
   public: std::vector<std::pair<Entity, math::Pose3d>> staticPoses;
 
-  /// \brief A variable that gets populated with poses. This also here as a
-  /// member variable to avoid repeated memory allocations and improve
-  /// performance.
-  public: msgs::Pose poseMsg;
-
-  /// \brief A variable that gets populated with poses. This also here as a
-  /// member variable to avoid repeated memory allocations and improve
-  /// performance.
-  public: msgs::Pose_V poseVMsg;
```

Comment:
This is definitely concerning. Do we know the root cause? I would be hesitant to merge this if we have to put this kind of condition on system authors.

Reply:
I've tried to reproduce the failure by reverting the changes to `PosePublisher` on this branch, running the [pose_publisher.sdf](https://github.com/gazebosim/gz-sim/blob/gz-sim8/examples/worlds/pose_publisher.sdf) example world, echoing the double pendulum pose topic, and then removing the pendulum, and I haven't seen any problems in testing on macOS. I suppose I could try to reproduce on Ubuntu

Reply:
From [a4c89e3](https://github.com/gazebosim/gz-sim/pull/2232/commits/a4c89e3ce4b6cf230ed6dd84649e0192e5d93e33):

> This commit removes an "optimization" in the PosePublisher that was
causing the system to segfault when an entity was removed. In reality
this commit is unrelated to the previous few commits. However, without this change
the previous change would cause removal of an entity to crash demos that
used pose publishers. I think the crux of the issue is that protobufs takeover
memory management of objects leading to a scenario where if a reference
was freed elsewhere, the protobuf would continue to hold said reference.
When we try to do operations on the protobuf we end up double-freeing
the memory and hence there is a 💥.

Any form of removal will require that system authors think about lifetimes. I could get the warehouse world to segfault on ubuntu when I made this change. Its been several months now so I havent tried recently.

Reply:
Yep can confirm without this change I get a crash in `PosePublisher`:
```
#11   Object "[0xffffffffffffffff]", at 0xffffffffffffffff, in 
#10   Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7fc6e332684f, in 
#9    Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x7fc6e3294ac2, in 
#8    Object "/lib/x86_64-linux-gnu/libstdc++.so.6", at 0x7fc6df4dc252, in 
#7    Object "/home/arjoc/ws/gz/ionic/install/lib/libgz-sim9.so.9", at 0x7fc6de77203e, in 
#6    Object "/home/arjoc/ws/gz/ionic/install/lib/gz-sim-9/plugins/libgz-sim-pose-publisher-system.so", at 0x7fc6aabfa3b7, in 
#5    Object "/home/arjoc/ws/gz/ionic/install/lib/gz-sim-9/plugins/libgz-sim-pose-publisher-system.so", at 0x7fc6aabf9ef0, in 
#4    Object "/home/arjoc/ws/gz/ionic/install/lib/libgz-transport14.so.14", at 0x7fc6de8ed6b0, in gz::transport::v14::Node::Publisher::Publish(google::protobuf::Message const&)
#3    Object "/home/arjoc/ws/gz/ionic/install/lib/libgz-msgs11.so.11", at 0x7fc6dde8e156, in gz::msgs::Pose::ByteSizeLong() const
#2    Object "/home/arjoc/ws/gz/ionic/install/lib/libgz-msgs11.so.11", at 0x7fc6dddf11e8, in gz::msgs::Header::ByteSizeLong() const
#1    Object "/home/arjoc/ws/gz/ionic/install/lib/libgz-msgs11.so.11", at 0x7fc6dddeec5a, in gz::msgs::Header_Map::ByteSizeLong() const
#0    Object "/lib/x86_64-linux-gnu/libprotobuf.so.23", at 0x7fc6dd8f3922, in google::protobuf::RepeatedPtrField<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::Get(int) const
Segmentation fault (Address not mapped to object [0x8])
```

Reply:
The stack trace indicates that the PosePublisher actually tried to publish after it was removed. To me, this looks like we haven't cleared the `PostUpdate` thread queue. Somehow we're calling the `PostUpdate` ona  deleted object? 

The `PosePublisher` system is not doing anything special. It's simply publishing in the `PostUpdate` call, so I don't think we can justify requiring system authors to think this deeply about lifetimes. I would understand having this requirement if it had its own thread like the `Sensors` system. But even then, the destructor of the system should be called, which should ideally terminate any internal threads and join them.

Reply:
You're right. I'll go figure it out. It seems like not all the threads are terminating cleanly. I dont also call the destructors the threads.

Reply:
maybe just make sure the `transport::Node` and `transport::Publisher` objects are defined at the end of the object so they get destroyed first?

~~~
diff --git a/src/systems/pose_publisher/PosePublisher.cc b/src/systems/pose_publisher/PosePublisher.cc
index 58661d481..56ea3e52b 100644
--- a/src/systems/pose_publisher/PosePublisher.cc
+++ b/src/systems/pose_publisher/PosePublisher.cc
@@ -83,18 +83,9 @@ class gz::sim::systems::PosePublisherPrivate
       const msgs::Time &_stampMsg,
       transport::Node::Publisher &_publisher);
 
-  /// \brief Gazebo communication node.
-  public: transport::Node node;
-
-  /// \brief publisher for pose data
-  public: transport::Node::Publisher posePub;
-
   /// \brief True to publish static transforms to a separate topic
   public: bool staticPosePublisher = false;
 
-  /// \brief publisher for pose data
-  public: transport::Node::Publisher poseStaticPub;
-
   /// \brief Model interface
   public: Model model{kNullEntity};
 
@@ -171,6 +162,15 @@ class gz::sim::systems::PosePublisherPrivate
 
   /// \brief Whether cache variables have been initialized
   public: bool initialized{false};
+
+  /// \brief Gazebo communication node.
+  public: transport::Node node;
+
+  /// \brief publisher for pose data
+  public: transport::Node::Publisher posePub;
+
+  /// \brief publisher for pose data
+  public: transport::Node::Publisher poseStaticPub;
 };
 
 //////////////////////////////////////////////////
~~~

Reply:
I think the issue is SystemContainer does not call the destructor in the first place.

Reply:
I was able to reproduce the crash with the following world, but my patch appears to fix it:

~~~
gz sim -v 4 "https://fuel.gazebosim.org/1.0/OpenRobotics/worlds/Tugbot in Warehouse"
~~~

Reply:
@azeey what do you think of @scpeter's solution?

Also do I need to call destructors upon deletion? Right now I just defer calling the destructor until the next instance of remove is called.

Reply:
Destructors are now called in [a764498](https://github.com/gazebosim/gz-sim/pull/2232/commits/a764498bfe23f5e468820aab75dcca5d5adba578).

Reply:
Based on print statements I can guarantee that the pose publisher is actually being terminated. @scpeters patch causes a crash still. I am more or less sure this has to do with abusing a ProtoBuf as a storage class. Specifically I ran the warehouse world. I despawned entity 16. The threads belonging to 16 cleanly exit:
![Screenshot from 2024-06-20 09-02-48](https://github.com/gazebosim/gz-sim/assets/542272/d727fafc-2e6b-4510-8feb-8bb1bae5ca29)

The thread that crashes belongs to entity 216 (which incidentally has a pose publisher as well): 

![Screenshot from 2024-06-20 08-53-18](https://github.com/gazebosim/gz-sim/assets/542272/2eb93072-a3f7-451e-967b-cefc6079688d)

Reply:
So I added some debug messages in the azeey/debug_remove_sytems branch and ran the `Tugbot in Warehouse` example at a slow update rate:
```
gz sim -z 10 "https://fuel.gazebosim.org/1.0/OpenRobotics/worlds/Tugbot in Warehouse"
```

Before removing the tugbot, here's the print out (repeated) from the PosePublisher. The format is `<thread id>: <entity>`. There are 6 PosePublishers each attached to a unique entity.
```
140267759843008:166
140267776628416:148
140267751450304:175
140267768235712:157
140268139423424:16
140267743057600:216

```

I then removed the tugbot and this is the print out right before crashing:

```
140267759843008:166
140267768235712:157
140267751450304:175
140267776628416:216
140267743057600:216
```
You can see that thread `140267776628416` is now associated with entity 216 instead of 148. This tells me there's some memory corruption happening.

Reply:
Thanks for catching this I'm investigating.

Reply:
883a0aba5 solves this.ll